### DISCUSSION TITLE
Expose Symbol.sealedStrictDescendants in Quoted API

### DISCUSSION THREAD

**Discussion by cchantep** [scala/scala3#17697] [Feature Requests]
Similar to previous Scala2 `knownDirectSubclasses` and based on dotc `sealedStrictDescendants`.

```patch
diff --git a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
index 3de7f0bfc6..0f69c849b0 100644
--- a/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
+++ b/compiler/src/scala/quoted/runtime/impl/QuotesImpl.scala
@@ -2622,6 +2622,9 @@ class QuotesImpl private (using val ctx: Context) extends Quotes, QuoteUnpickler
         def companionModule: Symbol = self.denot.companionModule
         def children: List[Symbol] = self.denot.children
 
+        def sealedStrictDescendants: List[Symbol] = 
+          self.denot.sealedStrictDescendants
+
         def show(using printer: Printer[Symbol]): String = printer.show(self)
 
       end extension
diff --git a/library/src/scala/quoted/Quotes.scala b/library/src/scala/quoted/Quotes.scala
index a375709844..4c2856e3b3 100644
--- a/library/src/scala/quoted/Quotes.scala
+++ b/library/src/scala/quoted/Quotes.scala
@@ -3783,6 +3783,10 @@ trait Quotes { self: runtime.QuoteUnpickler & runtime.QuoteMatching =>
 
         /** Case class or case object children of a sealed trait or cases of an `enum`. */
         def children: List[Symbol]
+
+        /** Recursively assemble all children of this symbol */
+        def sealedStrictDescendants: List[Symbol]
+
       end extension
     }
```

**Comment by bishabosha** [comment]
This could be implemented in the quotes API rather than by an intrinsic

**Comment by cchantep** [comment]
> This could be implemented in the quotes API rather than by an intrinsic

What do you mean?

The diff *is* an implementation suggestion in the quotes API based in underlying dotc.

**Comment by bishabosha** [comment]
I.e. my suggestion is to copy-paste the implementation from inside dotc and implement it in `scala.quoted.Quotes`, rather than leave it abstract (i.e. it is computed from recursively calling `children`). I acknowledge the one in the compiler is using some internal data structures - but its just a point to evaluate if its worth creating more dependencies

**Comment by cchantep** [comment]
I don't see any benefit from reimplementing/reinventing the logic which is already available in dotc.
Most of the `QuotesImpl` is done in this way, by "just" mapping.

**Comment by cchantep** [comment]
up

**Comment by cchantep** [comment]
up

**Comment by sjrd** [comment]
@cchantep Please stop posting "up" every week. It won't make it happen any faster, and it is only irritating the people who might do it.
